File: oo/josephus.py
"""
$ python josephus.py
26
object oriented
105.884628057
105.946704865
106.000247002
105.890681982
105.9042871
105.917474985
105.907362938
105.803264856
105.783190012
105.89677
"""

kth = 3; k = 1

class Soldier(object):

   def __init__(self,previousSoldier,number):

      self.number = number;

      if previousSoldier:
         self.next = previousSoldier.next
         previousSoldier.next = self
         self.previous = previousSoldier
         self.next.previous = self
      else:
         self.next = self
         self.previous = self


   def countoff(self):
      global k
      if k == kth:
         self.previous.next = self.next;
         self.next.previous = self.previous
         k = 1
      else:
         k += 1
      return self.next 


def countoffSoldiers(n,nth):
   global kth,k
   #kth = nth; k = 1 # standard Josephus problem
   kth = nth; k = nth

   soldier = None
   for i in (range(1,n+1)):
      soldier = Soldier(soldier,i)

   soldier = soldier.next

   while soldier.next != soldier:
      soldier = soldier.countoff()

   return soldier.number


#-----------------------------------------------------------------------------

import time
def run_iterations(iterations, times) :
    print "object oriented" 
    for t in range(times) :
        start = time.time()
        for i in range(iterations) :
            countoffSoldiers(40,3)
        end = time.time()
        print ((end - start) * 1000000 / iterations)

print countoffSoldiers(40,3)
ITER = 1000000
run_iterations(ITER,10)

File: oo/josephus.js
/*
$ /usr/local/src/v8/d8 josephus.js
26
object oriented
2.557
2.572
2.577
2.578
2.578
2.578
2.578
2.572
2.572
2.571
*/

function Soldier(previousSoldier,number){
   this.number = number

   if (previousSoldier != null){

      this.next = previousSoldier.next
      previousSoldier.next = this

      this.previous = previousSoldier
      this.next.previous = this

   } else {
      this.next = this
      this.previous = this
   }
}


Soldier.prototype.countoff = function(shout){
   if (shout.wasKth()) {

      this.previous.next = this.next
      this.next.previous = this.previous
   } 
   return this.next 
} 


function countoffSoldiers(n,kth) {
   var soldier = null	
   for (var i = 1; i <= n; i++) soldier = new Soldier(soldier,i) 

   soldier = soldier.next // set to the first Soldier constructed

   var shout = new Shoutout(kth) 

   while (soldier.next != soldier) 
      soldier = soldier.countoff(shout)

   return soldier.number
} 



//function Shoutout(n) { this.kth = n; this.k = 1 }   // standard Josephus problem
function Shoutout(n) { this.kth = n; this.k = n }

Shoutout.prototype.wasKth = function(){
   var wasKth = this.k == this.kth
   if (wasKth) this.k = 1; else this.k++;
   return wasKth
}


//-----------------------------------------------------------------------------

function run_iterations(iterations, times) {
    for (var t = 0 ; t < times ; t++) {
        var start = new Date()
        for(var i = 0 ; i < iterations ; i++) {
            countoffSoldiers(40,3)
        }
        var end = new Date()
        print ((end.getTime() - start.getTime()) * 1000 / ITERS)
    }
}

print(countoffSoldiers(40,3))
ITERS = 1000000
print ("object oriented")
run_iterations(ITERS,10)
